consensus/core/src/block.rs

```diff
@@ -294,13 +300,44 @@ pub(crate) struct SignedBlock {
 impl SignedBlock {
     // TODO: add verification.
 
-    // TODO: will refactor once the signing approach has been introduced.
-    pub(crate) fn new(block: Block) -> Self {
+    /// Should only be used when constructing the genesis blocks
+    pub(crate) fn new_genesis(block: Block) -> Self {
         Self {
             inner: block,
             signature: Bytes::default(),
         }
     }
+    pub(crate) fn new<S>(block: Block, signer: &S) -> ConsensusResult<Self>
+    where
+        S: Signer<NetworkKeySignature>,
+    {
+        let digest = compute_digest(&block)?;
+        let signature = NetworkKeySignature::new(&to_intent_message(digest), signer);
+        let signature_bytes = signature.as_bytes();
+        Ok(Self {
+            inner: block,
+            signature: signature_bytes.to_vec().into(),
+        })
+    }
+
+    #[allow(unused)]
+    /// This method is only verifying the block's signature for now. It does not do any kind of validation so it can
+    /// fail for numerous reasons.
+    pub(crate) fn verify(&self, context: &Context) -> ConsensusResult<()> {
+        let digest = compute_digest(&self.inner)?;
```

Comment:
Shall we revert to the previous approach of deriving the block digest via the `Block` it self? I think we use it in multiple places and maybe it makes less sense to live out of it?

Reply:
I would vote for reverting to the previous approach. However I am unsure of the benefits of having it as an external method. Will see what @mwtian says.

Reply:
> and should not be called often because it does not cache the result

agreed, that's why I am thinking of falling back to the previous approach of using the OnceCell could benefit here?

> is that it will make it harder to have a cached implementation for VerifiedBlock::digest()

well the VerifiedBlock it self doesn't need to cache and compute the digest at all. It can differ that completely to Block, can't it?